lib/engine-object-file/src/artifact.rs

```diff
@@ -0,0 +1,485 @@
+        let signatures = metadata
+            .compile_info
+            .module
+            .signatures
+            .values()
+            .map(|sig| signature_registry.register(sig))
+            .collect::<PrimaryMap<_, _>>();
+
+        Ok(Self {
+            metadata,
+            module_bytes,
+            finished_functions: finished_functions.into_boxed_slice(),
+            finished_dynamic_function_trampolines: finished_dynamic_function_trampolines
+                .into_boxed_slice(),
+            signatures: signatures.into_boxed_slice(),
+            metadata_length,
+        })
+    }
+
+    /// Compile a data buffer into a `ObjectFileArtifact`, which may then be instantiated.
+    #[cfg(not(feature = "compiler"))]
+    pub fn new(_engine: &ObjectFileEngine, _data: &[u8]) -> Result<Self, CompileError> {
+        Err(CompileError::Codegen(
+            "Compilation is not enabled in the engine".to_string(),
+        ))
+    }
+
+    /// Deserialize a `ObjectFileArtifact` from bytes.
+    ///
+    /// # Safety
+    ///
+    /// The bytes must represent a serialized WebAssembly module.
+    pub unsafe fn deserialize(
+        engine: &ObjectFileEngine,
+        bytes: &[u8],
+    ) -> Result<Self, DeserializeError> {
+        let mut reader = bytes;
+        let data_len = leb128::read::unsigned(&mut reader).unwrap() as usize;
+
+        let metadata: ModuleMetadata = bincode::deserialize(&bytes[10..(data_len + 10)]).unwrap();
+
+        const WORD_SIZE: usize = mem::size_of::<usize>();
+        let mut byte_buffer = [0u8; WORD_SIZE];
+
+        let mut cur_offset = data_len + 10;
+        for i in 0..WORD_SIZE {
+            byte_buffer[i] = bytes[cur_offset + i];
+        }
+        cur_offset += WORD_SIZE;
+
+        let num_finished_functions = usize::from_ne_bytes(byte_buffer);
+        let mut finished_functions = PrimaryMap::new();
+
+        #[repr(C)]
+        struct SlicePtr {
+            ptr: usize,
+            len: usize,
+        }
+
+        let mut engine_inner = engine.inner_mut();
+        let signature_registry = engine_inner.signatures();
+        let mut sig_map: BTreeMap<SignatureIndex, VMSharedSignatureIndex> = BTreeMap::new();
+
+        // read finished functions in order now...
+        for i in 0..num_finished_functions {
+            let sig_idx = metadata.compile_info.module.functions[FunctionIndex::new(i)];
+            let func_type = &metadata.compile_info.module.signatures[sig_idx];
+            let vm_shared_idx = signature_registry.register(&func_type);
+            sig_map.insert(sig_idx, vm_shared_idx);
+
+            let mut sp = SlicePtr { ptr: 0, len: 0 };
+            for j in 0..WORD_SIZE {
+                byte_buffer[j] = bytes[cur_offset + j];
+            }
+            sp.ptr = usize::from_ne_bytes(byte_buffer);
+            cur_offset += WORD_SIZE;
+            // REVIEW: we can also serialize and read back lengths, do we want to do this?
```

Comment:
What do you mean?

Reply:
This has to do with the FunctionBodyLengths... currently they're all 0 and everything works, I wasn't sure if we needed it

Reply:
It's apparently useful for debug output.  I'm going to not fix this in this PR (I spent some time trying to do so today and wasn't successful), but we can add it later -- I have comments in all the places that would need to be updated.